dynamic-check/src/objects/frame.rs

```diff
@@ -0,0 +1,1198 @@
+        }
+    }
+
+    ///
+    /// Returns a `FunctionType` given a function identifier.
+    ///
+    fn parse_program_function(&mut self, identifier: &Identifier, _span: &Span) -> Result<FunctionType, FrameError> {
+        self.user_defined_types
+            .get_function_type(&identifier.name)
+            .map(|function_type| function_type.to_owned())
+            .ok_or_else(|| FrameError::undefined_function(identifier))
+    }
+
+    ///
+    /// Returns a `CircuitFunctionType` given a circuit expression and function identifier.
+    ///
+    fn parse_circuit_function_type(
+        &mut self,
+        expression: &Expression,
+        identifier: &Identifier,
+        span: &Span,
+    ) -> Result<&CircuitFunctionType, FrameError> {
+        // Parse circuit name.
+        let type_ = self.parse_expression(expression)?;
+
+        // Get circuit type.
+        let circuit_type = self.parse_circuit_name(type_, span)?;
+
+        // Find circuit function by identifier.
+        circuit_type
+            .member_function_type(identifier)
+            .ok_or_else(|| FrameError::undefined_circuit_function(identifier))
+    }
+
+    ///
+    /// Returns a `FunctionType` given a circuit expression and non-static function identifier.
+    ///
+    fn parse_circuit_function(
+        &mut self,
+        expression: &Expression,
+        identifier: &Identifier,
+        span: &Span,
+    ) -> Result<FunctionType, FrameError> {
+        // Find circuit function type.
+        let circuit_function_type = self.parse_circuit_function_type(expression, identifier, span)?;
+
+        // Check that the function is non-static.
+        if circuit_function_type.attributes.contains(&Attribute::Static) {
+            return Err(FrameError::invalid_static_access(identifier));
+        }
+
+        // Return the function type.
+        Ok(circuit_function_type.function.to_owned())
+    }
+
+    ///
+    /// Returns a `FunctionType` given a circuit expression and static function identifier.
+    ///
+    fn parse_static_circuit_function(
+        &mut self,
+        expression: &Expression,
+        identifier: &Identifier,
+        span: &Span,
+    ) -> Result<FunctionType, FrameError> {
+        // Find circuit function type.
+        let circuit_function_type = self.parse_circuit_function_type(expression, identifier, span)?;
+
+        // Check that the function is static.
+        if !circuit_function_type.attributes.contains(&Attribute::Static) {
+            return Err(FrameError::invalid_member_access(identifier));
+        }
+
+        Ok(circuit_function_type.function.to_owned())
+    }
+
+    ///
+    /// Returns the type returned by calling the function.
+    ///
+    /// Does not attempt to evaluate the function call. We are just checking types at this step.
+    ///
+    fn parse_function_call(
+        &mut self,
+        expression: &Expression,
+        inputs: &Vec<Expression>,
```

Comment:
```suggestion
        inputs: impl ExactSizeIterator<Item = Expression>,
```
if possible; otherwise, `&[Expression]` should be used here